--- a/malloc/malloc.c	2020-08-05 05:17:00.000000000 +0300
+++ b/malloc/malloc.c	2020-12-22 21:17:55.000000000 +0300
@@ -1042,8 +1042,15 @@ static void*   memalign_check(size_t ali
 # define MAP_NORESERVE 0
 #endif
 
+#if defined(__linux__) && !defined(MAP_UNINITIALIZED)
+# include <asm-generic/mman-common.h>
+#endif
+#ifndef MAP_UNINITIALIZED
+# define MAP_UNINITIALIZED 0
+#endif
+
 #define MMAP(addr, size, prot, flags) \
- __mmap((addr), (size), (prot), (flags)|MAP_ANONYMOUS|MAP_PRIVATE, -1, 0)
+ __mmap((addr), (size), (prot), (flags)|MAP_ANONYMOUS|MAP_PRIVATE|MAP_UNINITIALIZED, -1, 0)
 
 
 /*
