--- a/drivers/video/via/via-core.c	2010-10-20 23:30:22.000000000 +0300
+++ b/drivers/video/via/via-core.c	2010-11-22 01:42:33.299998366 +0200
@@ -2,6 +2,7 @@
  * Copyright 1998-2009 VIA Technologies, Inc. All Rights Reserved.
  * Copyright 2001-2008 S3 Graphics, Inc. All Rights Reserved.
  * Copyright 2009 Jonathan Corbet <corbet@lwn.net>
+ * Copyright 2010 Dzianis Kahanovich <mahatma@eu.by>
  */
 
 /*
@@ -559,6 +560,58 @@ static void via_teardown_subdevs(void)
 		}
 }
 
+static void via_i2c_detect(void){
+	const u8 *edid;
+	struct fb_var_screeninfo var;
+	int i, n = 0, d;
+	struct i2c_adapter *adapter;
+
+	d = viafb_platform_epia_dvi == STATE_OFF &&
+		viafb_display_hardware_layout == HW_LAYOUT_LCD_DVI &&
+		viafb_DeviceStatus == CRT_Device &&
+		viafb_lcd_mode == LCD_OPENLDI &&
+		viafb_CRT_ON == STATE_ON &&
+		viafb_LCD_ON == STATE_OFF &&
+		viafb_DVI_ON == STATE_OFF &&
+		viafb_LCD2_ON == STATE_OFF &&
+		viafb_primary_dev == None_Device;
+	for (i = 0; i < VIAFB_NUM_PORTS; i++) {
+		adapter = viafb_find_i2c_adapter(i);
+		if (!adapter || !adapter->algo_data ||
+			! (edid = fb_ddc_read(adapter)))
+				continue;
+		if (fb_parse_edid(edid, &var))
+			goto free_edid;
+		if(n) {
+			fb_edid_to_monspecs(edid, &viafbinfo1->monspecs);
+			viaparinfo->lvds_setting_info2->lcd_panel_hres = var.xres;
+			viaparinfo->lvds_setting_info2->lcd_panel_vres = var.yres;
+			if (d) {
+				viafb_SAMM_ON = STATE_ON;
+				if (viafbinfo1->monspecs.input & FB_DISP_DDI)
+					viafb_LCD2_ON = STATE_ON;
+			}
+		} else {
+			fb_edid_to_monspecs(edid, &viafbinfo->monspecs);
+			viaparinfo->lvds_setting_info->lcd_panel_hres = var.xres;
+			viaparinfo->lvds_setting_info->lcd_panel_vres = var.yres;
+			if (d) {
+				viafb_DVI_ON = viaparinfo->tmds_setting_info->max_hres ? STATE_ON : STATE_OFF;
+				if (viafbinfo->monspecs.input & FB_DISP_DDI) {
+					viafb_DeviceStatus = LCD_Device;
+					viafb_primary_dev = LCD_Device;
+					viafb_LCD_ON = STATE_ON;
+				}
+			}
+		}
+		n++;
+free_edid:
+		kfree(edid);
+		if (n > 1)
+			break;
+	}
+}
+
 
 static int __devinit via_pci_probe(struct pci_dev *pdev,
 		const struct pci_device_id *ent)
@@ -591,6 +644,7 @@ static int __devinit via_pci_probe(struc
 	ret = via_fb_pci_probe(&global_dev);
 	if (ret)
 		goto out_subdevs;
+	via_i2c_detect();
 	return 0;
 
 out_subdevs:
