Separated to vendors-caches-4.8.patch to 4.9 compats
--- a/gcc/config/i386/driver-i386.c	2013-04-27 14:57:09.000000000 +0300
+++ b/gcc/config/i386/driver-i386.c	2013-04-27 15:04:32.859793073 +0300
@@ -391,7 +391,7 @@ const char *host_detect_local_cpu (int a
   unsigned int has_rdseed = 0, has_prfchw = 0, has_adx = 0;
   unsigned int has_osxsave = 0, has_fxsr = 0, has_xsave = 0, has_xsaveopt = 0;
 
-  bool arch;
+  bool arch;
 
   unsigned int l2sizekb = 0;
 
@@ -560,6 +561,32 @@ const char *host_detect_local_cpu (int a
       else
 	processor = PROCESSOR_PENTIUM;
     }
+  else if (vendor == signature_CENTAUR_ebx)
+     if (family == 6)
+	if (has_longmode)
+	  cpu = "core2";
+	else if (model > 9)
+	  cpu = arch ? "prescott" : "core2";
+	else if (model == 9)
+	  cpu = "c3-2";
+	else if (model >= 6)
+	  cpu = "c3";
+	else
+#ifdef PROCESSOR_GENERIC32
+	  processor = PROCESSOR_GENERIC32;
+#else
+	  processor = PROCESSOR_GENERIC;
+#endif
+     else if (has_3dnow)
+	cpu = "winchip2";
+     else if (has_mmx)
+	cpu = "winchip-c6";
+     else
+#ifdef PROCESSOR_GENERIC32
+	processor = PROCESSOR_GENERIC32;
+#else
+	processor = PROCESSOR_GENERIC;
+#endif
   else
     {
       switch (family)
