--- a/gcc/config/i386/i386.c	2009-10-09 15:06:20.000000000 +0300
+++ c/gcc/config/i386/i386.c	2009-10-09 16:55:52.000000000 +0300
@@ -3238,6 +3238,11 @@ override_options (bool main_args_p)
   /* Set the default value for -mstackrealign.  */
   if (ix86_force_align_arg_pointer == -1)
     ix86_force_align_arg_pointer = STACK_REALIGN_DEFAULT;
+  else if (!ix86_force_align_arg_pointer && (flag_tree_vectorize || optimize > 2) && STACK_REALIGN_DEFAULT) {
+    ix86_isa_flags &= ~OPTION_MASK_ISA_SSE_UNSET;
+    ix86_isa_flags_explicit |= OPTION_MASK_ISA_SSE_UNSET;
+    warning (0, "-mno-stackrealign present with no SSE4a cpu and -ftree-vectorize, SSE disabled");
+  }
 
   /* Validate -mincoming-stack-boundary= value or default it to
      MIN_STACK_BOUNDARY/PREFERRED_STACK_BOUNDARY.  */
--- a/gcc/config/i386/i386.h	2009-10-09 15:06:20.000000000 +0300
+++ c/gcc/config/i386/i386.h	2009-10-09 16:55:52.000000000 +0300
@@ -709,1 +709,3 @@ enum target_cpu_default
-#define STACK_REALIGN_DEFAULT 0
+/* "misalignsse" feature found near SSE4a, this way is simpiest */
+#define STACK_REALIGN_DEFAULT \
+	(((TARGET_SSE && !TARGET_SSE4A) && !TARGET_64BIT) ? 1 : 0)
--- a/gcc/doc/invoke.texi	2009-10-09 14:22:17.000000000 +0300
+++ c/gcc/doc/invoke.texi	2009-10-09 16:56:25.000000000 +0300
@@ -11898,6 +11898,7 @@ runtime stack if necessary.  This suppor
 a 4-byte aligned stack with modern codes that keep a 16-byte stack for
 SSE compatibility.  See also the attribute @code{force_align_arg_pointer},
 applicable to individual functions.
+It is enabled by default at SSE and !SSE4a on 32bit.
 
 @item -mpreferred-stack-boundary=@var{num}
 @opindex mpreferred-stack-boundary
--- a/gcc/config/i386/t-i386	2010-05-13 19:36:13.000000000 +0300
+++ c/gcc/config/i386/t-i386	2010-05-13 19:37:05.000000000 +0300
@@ -17,2 +17,4 @@
 # <http://www.gnu.org/licenses/>.
 
+TARGET_LIBGCC2_CFLAGS += -mno-stackrealign
+
