--- a/etc/xdg/ya/tint2rc	2016-09-02 16:41:19.578204278 +0300
+++ b/etc/xdg/ya/tint2rc	2016-09-02 16:42:52.635199599 +0300
@@ -16,9 +16,9 @@ border_color = #ffffff 30
 
 # Background 3: active tasks
 rounded = 3
-border_width = 0
-background_color = #ffffff 20
-border_color = #ffffff 40
+border_width = 1
+background_color = #ffffff 40
+border_color = #ffffff 48
 
 # Background 4: urgent tasks
 rounded = 3
