--- ProxyChannel.cxx.1
+++ ProxyChannel.cxx
@@ -965,7 +965,7 @@
 	PTRACE(tlevel, pstrm);
 }
 #else
-inline void PrintQ931(int, const char *, const char *, const Q931 *, const H225_H323_UserInformation *)
+void PrintQ931(int, const char *, const char *, const Q931 *, const H225_H323_UserInformation *)
 {
 	// nothing to do
 }
--- gk.cxx.1
+++ gk.cxx
@@ -123,7 +123,9 @@
 
 void ReloadHandler()
 {
+#if PTRACING
 	Gatekeeper::ReopenLogFile();
+#endif
 
 	// only one thread must do this
 	if (ReloadMutex.WillBlock())
@@ -162,1 +164,3 @@
+#if PTRACING
 	Gatekeeper::EnableLogFileRotation();
+#endif
@@ -600,1 +604,3 @@
+#if PTRACING
 	EnableLogFileRotation();
+#endif
--- gkauth.cxx.1
+++ gkauth.cxx
@@ -862,8 +862,8 @@
 	}
 
 	if (mechanisms.GetSize() > 0 && algorithmOIDs.GetSize() > 0) {
-		if (PTrace::CanTrace(4)) {
 #if PTRACING
+		if (PTrace::CanTrace(4)) {
 			ostream& strm = PTrace::Begin(4,__FILE__,__LINE__);
 			strm <<"GkAuth\tH.235 capabilities selected for GCF:\n";
 			strm <<"\tAuthentication mechanisms: \n";
@@ -873,8 +873,8 @@
 			for (i = 0; i < algorithmOIDs.GetSize(); i++)
 				strm << "\t\t" << algorithmOIDs[i] << '\n';
 			PTrace::End(strm);
-#endif
 		}
+#endif
 #ifdef OpenH323Factory
 	}
 #else
--- name.h.1
+++ name.h
@@ -20,7 +20,7 @@
 #ifndef NAME_H
 #define NAME_H "@(#) $Id: name.h,v 1.8 2009/02/09 15:26:21 willamowius Exp $"
 
-#if PTRACING
+#if 1
 
 class NamedObject 
 {
--- GkStatus.cxx.1	2010-10-22 14:08:13.000000000 +0300
+++ GkStatus.cxx	2010-10-22 14:09:10.000000000 +0300
@@ -974,6 +974,7 @@ void StatusClient::DoDebug(
 			);
 	} else {
 		if (args[1] *= "trc") {
+#if PTRACING
 			if(args.GetSize() >= 3) {
 				if((args[2] == "-") && (PTrace::GetLevel() > 0)) 
 					PTrace::SetLevel(PTrace::GetLevel()-1);
@@ -982,6 +983,7 @@ void StatusClient::DoDebug(
 				else PTrace::SetLevel(args[2].AsInteger());
 			}
 			WriteString(PString(PString::Printf, "Trace Level is now %d\r\n", PTrace::GetLevel()));
+#endif
 		} else if (args[1] *= "cfg") {
 			if (args.GetSize()>=4)
 				WriteString(GkConfig()->GetString(args[2],args[3],"") + "\r\n;\r\n");
