--- a/src/pages.c	2019-08-05 23:02:00.000000000 +0300
+++ b/src/pages.c	2020-12-22 21:10:09.320952788 +0300
@@ -15,6 +15,13 @@
 #endif
 #endif
 
+#if defined(__linux__) && !defined(MAP_UNINITIALIZED)
+#include <asm-generic/mman-common.h>
+#endif
+#ifndef MAP_UNINITIALIZED
+#define MAP_UNINITIALIZED 0
+#endif
+
 /******************************************************************************/
 /* Data. */
 
@@ -607,7 +614,7 @@ pages_boot(void) {
 	}
 
 #ifndef _WIN32
-	mmap_flags = MAP_PRIVATE | MAP_ANON;
+	mmap_flags = MAP_PRIVATE | MAP_ANON | MAP_UNINITIALIZED;
 #endif
 
 #ifdef JEMALLOC_SYSCTL_VM_OVERCOMMIT
