diff -pruN a/gcc/config/i386/i386.c c/gcc/config/i386/i386.c
--- a/gcc/config/i386/i386.c	2009-09-25 18:56:33.000000000 +0300
+++ c/gcc/config/i386/i386.c	2009-10-09 17:03:44.000000000 +0300
@@ -3090,6 +3090,11 @@ override_options (bool main_args_p)
   /* Set the default value for -mstackrealign.  */
   if (ix86_force_align_arg_pointer == -1)
     ix86_force_align_arg_pointer = STACK_REALIGN_DEFAULT;
+  else if (!ix86_force_align_arg_pointer && (flag_tree_vectorize || optimize > 2) && STACK_REALIGN_DEFAULT) {
+    ix86_isa_flags &= ~OPTION_MASK_ISA_SSE_UNSET;
+    ix86_isa_flags_explicit |= OPTION_MASK_ISA_SSE_UNSET;
+    warning (0, "-mno-stackrealign present with no SSE4a cpu and -ftree-vectorize, SSE disabled");
+  }
 
   /* Validate -mincoming-stack-boundary= value or default it to
      MIN_STACK_BOUNDARY/PREFERRED_STACK_BOUNDARY.  */
diff -pruN a/gcc/config/i386/i386.h c/gcc/config/i386/i386.h
--- a/gcc/config/i386/i386.h	2009-09-25 18:56:33.000000000 +0300
+++ c/gcc/config/i386/i386.h	2009-10-09 17:03:44.000000000 +0300
@@ -682,7 +682,9 @@ enum target_cpu_default
    need a 16-byte aligned stack, as required by SSE instructions.  If
    STACK_REALIGN_DEFAULT is 1 and PREFERRED_STACK_BOUNDARY_DEFAULT is
    128, stacks for all functions may be realigned.  */
-#define STACK_REALIGN_DEFAULT 0
+/* "misalignsse" feature found near SSE4a, this way is simpiest */
+#define STACK_REALIGN_DEFAULT \
+	(((TARGET_SSE && !TARGET_SSE4A) && !TARGET_64BIT) ? 1 : 0)
 
 /* Boundary (in *bits*) on which the incoming stack is aligned.  */
 #define INCOMING_STACK_BOUNDARY ix86_incoming_stack_boundary
diff -pruN a/gcc/config/i386/t-i386 c/gcc/config/i386/t-i386
--- a/gcc/config/i386/t-i386	2009-09-25 18:56:33.000000000 +0300
+++ c/gcc/config/i386/t-i386	2009-10-09 17:03:44.000000000 +0300
@@ -11,3 +11,5 @@ i386-c.o: $(srcdir)/config/i386/i386-c.c
   $(TM_H) $(RTL_H) $(TREE_H) $(TM_P_H) $(FLAGS_H) $(C_COMMON_H) $(GGC_H) \
   $(TARGET_H) $(TARGET_DEF_H) $(CPPLIB_H) $(C_PRAGMA_H)
 	$(CC) -c $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) $(srcdir)/config/i386/i386-c.c
+
+TARGET_LIBGCC2_CFLAGS += -mno-stackrealign
diff -pruN a/gcc/doc/invoke.texi c/gcc/doc/invoke.texi
--- a/gcc/doc/invoke.texi	2009-09-25 15:35:20.000000000 +0300
+++ c/gcc/doc/invoke.texi	2009-10-09 17:03:03.000000000 +0300
@@ -11252,6 +11252,7 @@ runtime stack if necessary.  This suppor
 a 4-byte aligned stack with modern codes that keep a 16-byte stack for
 SSE compatibility.  See also the attribute @code{force_align_arg_pointer},
 applicable to individual functions.
+It is enabled by default at SSE and !SSE4a on 32bit.
 
 @item -mpreferred-stack-boundary=@var{num}
 @opindex mpreferred-stack-boundary
