--- a/gcc/config/i386/i386.c	2014-06-17 13:27:04.637573878 +0300
+++ b/gcc/config/i386/i386.c	2014-06-17 13:37:57.733576055 +0300
@@ -3918,6 +3918,11 @@ ix86_option_override_internal (bool main
   /* Set the default value for -mstackrealign.  */
   if (opts->x_ix86_force_align_arg_pointer == -1)
     opts->x_ix86_force_align_arg_pointer = STACK_REALIGN_DEFAULT;
+  else if (STACK_REALIGN_DEFAULT && !opts->x_ix86_force_align_arg_pointer && (flag_tree_loop_vectorize || flag_tree_slp_vectorize || optimize > 2)) {
+    opts->x_ix86_isa_flags &= ~(OPTION_MASK_ISA_SSE | OPTION_MASK_ISA_SSE2 | OPTION_MASK_ISA_SSE3 | OPTION_MASK_ISA_SSSE3); // more?
+    opts->x_ix86_isa_flags_explicit |= (OPTION_MASK_ISA_SSE | OPTION_MASK_ISA_SSE2 | OPTION_MASK_ISA_SSE3 | OPTION_MASK_ISA_SSSE3);
+    warning (0, "-mno-stackrealign present with no SSE4a cpu and -ftree-vectorize, SSE disabled");
+  }
 
   ix86_default_incoming_stack_boundary = PREFERRED_STACK_BOUNDARY;
 
--- a/gcc/config/i386/i386.h	2014-06-17 13:27:05.205580849 +0300
+++ b/gcc/config/i386/i386.h	2014-06-17 13:28:57.702961096 +0300
@@ -748,1 +748,3 @@ extern const char *host_detect_local_cpu
-#define STACK_REALIGN_DEFAULT 0
+/* "misalignsse" feature found near SSE4a, this way is simpiest */
+#define STACK_REALIGN_DEFAULT \
+	(((TARGET_SSE && !TARGET_SSE4A) && !TARGET_64BIT) ? 1 : 0)
--- a/gcc/config/i386/t-i386	2014-03-05 12:52:18.000000000 +0300
+++ b/gcc/config/i386/t-i386	2014-06-17 13:28:57.706961146 +0300
@@ -17,2 +17,4 @@
 # <http://www.gnu.org/licenses/>.
 
+TARGET_LIBGCC2_CFLAGS += -mno-stackrealign
+
--- a/gcc/doc/invoke.texi	2014-06-17 13:27:05.129579915 +0300
+++ b/gcc/doc/invoke.texi	2014-06-17 13:28:57.706961146 +0300
@@ -15200,6 +15200,7 @@ run-time stack if necessary.  This suppo
 4-byte stack alignment with modern codes that keep 16-byte stack alignment for
 SSE compatibility.  See also the attribute @code{force_align_arg_pointer},
 applicable to individual functions.
+It is enabled by default at SSE and !SSE4a on 32bit.
 
 @item -mpreferred-stack-boundary=@var{num}
 @opindex mpreferred-stack-boundary
