--- a/Makefile.pre.in	2017-03-12 15:08:48.000000000 +0300
+++ b/Makefile.pre.in	2017-03-12 15:19:47.734421143 +0300
@@ -1409,6 +1409,7 @@ profile-removal:
 	find . -name '*.gc??' -exec rm -f {} ';'
 	find . -name '*.profclang?' -exec rm -f {} ';'
 	find . -name '*.dyn' -exec rm -f {} ';'
+	true
 
 clobber: clean profile-removal
 	-rm -f $(BUILDPYTHON) $(PGEN) $(LIBRARY) $(LDLIBRARY) $(DLLLIBRARY) \
