--- a/src/wayland-client.h	2012-02-09 17:37:08.000000000 +0300
+++ b/src/wayland-client.h	2012-03-26 19:59:28.000000000 +0300
@@ -72,6 +72,7 @@ typedef void (*wl_callback_func_t)(void
 
 struct wl_display *wl_display_connect(const char *name);
 void wl_display_destroy(struct wl_display *display);
+#define wl_display_disconnect wl_display_destroy
 int wl_display_get_fd(struct wl_display *display,
 		      wl_display_update_func_t update, void *data);
 void wl_display_iterate(struct wl_display *display, uint32_t mask);
