--- a/genkernel	2015-08-19 20:58:58.000000000 +0300
+++ b/genkernel	2015-08-19 21:21:35.543748306 +0300
@@ -27,8 +27,7 @@
 # /etc/genkernel.conf if nobody has specified one.
 
 # NOTE: We are look for --config=... in a way that doesn't modify $@ since we access that again, later
-for (( i=1; i<=$# ; i=i+1 )); do
-	eval arg="\$$i"
+for arg in "${@}"; do
 	[[ "${arg}" = --config=* ]] && CMD_GK_CONFIG=`parse_opt "${arg}"`
 done
 
