diff -pruN a/include/net/netfilter/nf_conntrack.h b0/include/net/netfilter/nf_conntrack.h
--- a/include/net/netfilter/nf_conntrack.h	2008-10-29 17:59:26.000000000 +0200
+++ b0/include/net/netfilter/nf_conntrack.h	2008-11-17 16:16:14.000000000 +0200
@@ -109,6 +109,9 @@ struct nf_conn
 
 	/* Timer function; drops refcnt when it goes off. */
 	struct timer_list timeout;
+#ifdef CONFIG_NETFILTER_XT_MATCH_CONNLIMIT_ACTIVE
+	unsigned long use_time;
+#endif
 
 #if defined(CONFIG_NF_CONNTRACK_MARK)
 	u_int32_t mark;
diff -pruN a/net/netfilter/Kconfig b0/net/netfilter/Kconfig
--- a/net/netfilter/Kconfig	2008-11-01 13:45:55.000000000 +0200
+++ b0/net/netfilter/Kconfig	2008-11-17 16:16:14.000000000 +0200
@@ -525,6 +525,15 @@ config NETFILTER_XT_MATCH_CONNLIMIT
 	  This match allows you to match against the number of parallel
 	  connections to a server per client IP address (or address block).
 
+config NETFILTER_XT_MATCH_CONNLIMIT_ACTIVE
+	bool '"connlimit" only active connections'
+	depends on NETFILTER_XT_MATCH_CONNLIMIT
+	help
+	  Skip keepalive and other not closed inactive connections.
+
+	  N to care memory usage,
+	  Y to care channel usage.
+
 config NETFILTER_XT_MATCH_CONNMARK
 	tristate  '"connmark" connection mark match support'
 	depends on NF_CONNTRACK
diff -pruN a/net/netfilter/nf_conntrack_core.c b0/net/netfilter/nf_conntrack_core.c
--- a/net/netfilter/nf_conntrack_core.c	2008-11-01 13:45:55.000000000 +0200
+++ b0/net/netfilter/nf_conntrack_core.c	2008-11-17 16:16:14.000000000 +0200
@@ -367,6 +367,9 @@ __nf_conntrack_confirm(struct sk_buff *s
 	/* Timer relative to confirmation time, not original
 	   setting time, otherwise we'd get timer wrap in
 	   weird delay cases. */
+#ifdef CONFIG_NETFILTER_XT_MATCH_CONNLIMIT_ACTIVE
+	ct->use_time = jiffies;
+#endif
 	ct->timeout.expires += jiffies;
 	add_timer(&ct->timeout);
 	atomic_inc(&ct->ct_general.use);
@@ -840,6 +843,10 @@ void __nf_ct_refresh_acct(struct nf_conn
 
 	spin_lock_bh(&nf_conntrack_lock);
 
+#ifdef CONFIG_NETFILTER_XT_MATCH_CONNLIMIT_ACTIVE
+	ct->use_time = jiffies;
+#endif
+
 	/* Only update if this is not a fixed timeout */
 	if (test_bit(IPS_FIXED_TIMEOUT_BIT, &ct->status))
 		goto acct;
diff -pruN a/net/netfilter/xt_connlimit.c b0/net/netfilter/xt_connlimit.c
--- a/net/netfilter/xt_connlimit.c	2008-10-29 17:59:37.000000000 +0200
+++ b0/net/netfilter/xt_connlimit.c	2008-11-17 16:25:00.000000000 +0200
@@ -29,6 +29,11 @@
 #include <net/netfilter/nf_conntrack_core.h>
 #include <net/netfilter/nf_conntrack_tuple.h>
 
+#ifdef CONFIG_NETFILTER_XT_MATCH_CONNLIMIT_ACTIVE
+/* 10 sec */
+#define CONNLIMIT_TIMEOUT (10*HZ)
+#endif
+
 /* we will save the tuples of all connections we care about */
 struct xt_connlimit_conn {
 	struct list_head list;
@@ -103,7 +108,8 @@ static int count_them(struct xt_connlimi
 		      const struct nf_conntrack_tuple *tuple,
 		      const union nf_inet_addr *addr,
 		      const union nf_inet_addr *mask,
-		      const struct xt_match *match)
+		      const struct xt_match *match,
+		      const struct nf_conn *ct)
 {
 	const struct nf_conntrack_tuple_hash *found;
 	struct xt_connlimit_conn *conn;
@@ -157,9 +163,20 @@ static int count_them(struct xt_connlimi
 		}
 
 		if (same_source_net(addr, mask, &conn->tuple.src.u3,
-		    match->family))
+		    match->family)) {
+#ifdef CONFIG_NETFILTER_XT_MATCH_CONNLIMIT_ACTIVE
+			if (found_ct != NULL) {
+				/* count only active connections */
+				if (jiffies - found_ct->use_time > CONNLIMIT_TIMEOUT)
+				    continue;
+				/* force to first connections to be active */
+				if (found_ct == ct)
+				    break;
+			}
+#endif
 			/* same source network -> be counted! */
 			++matches;
+		}
 	}
 
 	rcu_read_unlock();
@@ -205,7 +222,7 @@ connlimit_mt(const struct sk_buff *skb, 
 
 	spin_lock_bh(&info->data->lock);
 	connections = count_them(info->data, tuple_ptr, &addr,
-	                         &info->mask, par->match);
+	                         &info->mask, par->match, ct);
 	spin_unlock_bh(&info->data->lock);
 
 	if (connections < 0) {
