--- mcu.cxx.1	2010-10-22 12:48:22.273501000 +0300
+++ mcu.cxx	2010-10-22 12:49:27.763501001 +0300
@@ -309,12 +309,6 @@ PBoolean OpenMCU::Initialise(const char
   // Create invite conference page
   httpNameSpace.AddResource(new InvitePage(*this, authority), PHTTPSpace::Overwrite);
 
-  // Add log file links
-  if (!systemLogFileName && (systemLogFileName != "-")) {
-    httpNameSpace.AddResource(new PHTTPFile("logfile.txt", systemLogFileName, authority));
-    httpNameSpace.AddResource(new PHTTPTailFile("tail_logfile", systemLogFileName, authority));
-  }
-
   //  create the home page
   static const char welcomeHtml[] = "welcome.html";
   if (PFile::Exists(welcomeHtml))
@@ -333,12 +327,6 @@ PBoolean OpenMCU::Initialise(const char
          << PHTML::HotLink("Invite") << "Invite user to conference" << PHTML::HotLink()
          << PHTML::Paragraph();
 
-    if (!systemLogFileName && systemLogFileName != "-")
-      html << PHTML::HotLink("logfile.txt") << "Full Log File" << PHTML::HotLink()
-           << PHTML::BreakLine()
-           << PHTML::HotLink("tail_logfile") << "Tail Log File" << PHTML::HotLink()
-           << PHTML::Paragraph();
- 
     html << PHTML::HRule()
          << GetCopyrightText()
          << PHTML::Body();
