--- src/h323pluginmgr.cxx.1	2010-10-22 13:16:43.633501002 +0300
+++ src/h323pluginmgr.cxx	2010-05-25 15:54:59.000000000 +0300
@@ -1775,7 +1775,9 @@ bool OpalPluginCodec::SetMediaFormat(Opa
     PopulateMediaFormatOptions(codecDefn, fmt);
     PopulateMediaFormatFromGenericData(fmt, 
           (PluginCodec_H323GenericCodecData *)codecDefn->h323CapabilityData);
+#if PTRACING
     OpalMediaFormat::DebugOptionList(fmt);
+#endif
     return true;
 }
 
--- src/rtp.cxx.1	2010-10-22 13:08:55.993501001 +0300
+++ src/rtp.cxx	2010-10-22 13:09:24.723501001 +0300
@@ -1401,8 +1401,8 @@ RTP_Session::SendReceiveStatus RTP_Sessi
 }
 
 
-void RTP_Session::OnRxSenderReport(const SenderReport & PTRACE_PARAM(sender),
-                                   const ReceiverReportArray & PTRACE_PARAM(reports))
+void RTP_Session::OnRxSenderReport(const SenderReport & sender,
+                                   const ReceiverReportArray & reports)
 {
    userData->OnRxSenderReport(sessionID,sender,reports);
 
