Separated to vendors-caches-4.8.patch to 4.9 compats
--- a/gcc/config/i386/driver-i386.c	2013-04-27 14:53:15.376967086 +0300
+++ b/gcc/config/i386/driver-i386.c	2013-04-27 15:05:31.890844461 +0300
@@ -391,7 +391,7 @@ const char *host_detect_local_cpu (int a
   unsigned int has_rdseed = 0, has_prfchw = 0, has_adx = 0;
   unsigned int has_osxsave = 0, has_fxsr = 0, has_xsave = 0, has_xsaveopt = 0;
 
-  bool arch, tune, astune;
+  bool arch, tune, astune;
 
   unsigned int l2sizekb = 0;
 
@@ -562,6 +563,24 @@ const char *host_detect_local_cpu (int a
       else
 	processor = PROCESSOR_PENTIUM;
     }
+  else if (vendor == signature_CENTAUR_ebx)
+     if (family == 6)
+	if (has_longmode)
+	  cpu = "core2";
+	else if (model > 9)
+	  cpu = arch ? "prescott" : "core2";
+	else if (model == 9)
+	  cpu = astune ? "generic32" : "c3-2";
+	else if (model >= 6)
+	  cpu = astune ? "generic32" : "c3";
+	else
+	  processor = PROCESSOR_GENERIC32;
+     else if (has_3dnow)
+	cpu = astune ? "generic32" : "winchip2";
+     else if (has_mmx)
+	cpu = astune ? "generic32" : "winchip-c6";
+     else
+	processor = PROCESSOR_GENERIC32;
   else
     {
       switch (family)
