Rare combinations bug, but must be neutral
--- a/include/misc.h	2015-03-06 02:45:05.000000000 +0300
+++ b/include/misc.h	2015-03-06 02:43:37.000000000 +0300
@@ -108,6 +108,7 @@ typedef struct _CallbackList *CallbackLi
 
 typedef struct _xReq *xReqPtr;
 
+#include "xorg-server.h"
 #include "os.h"                 /* for ALLOCATE_LOCAL and DEALLOCATE_LOCAL */
 #include <X11/Xfuncs.h>         /* for bcopy, bzero, and bcmp */
 
