diff -pruN a/openivr/main.cxx b/openivr/main.cxx
--- a/openivr/main.cxx	2011-07-17 13:17:33.000000000 +0300
+++ b/openivr/main.cxx	2011-08-24 14:06:58.000000000 +0300
@@ -156,10 +156,12 @@ PBoolean OpenIVR::Initialise(const char
 
   // Set log level as early as possible
   SetLogLevel((PSystemLog::Level)cfg.GetInteger(LogLevelKey, GetLogLevel()));
+#if PTRACING
   if (GetLogLevel() >= PSystemLog::Warning)
     PTrace::SetLevel(GetLogLevel()-PSystemLog::Warning);
   else
     PTrace::SetLevel(0);
+#endif
 
   PString userName = cfg.GetString(UsernameKey, "admin");
   PString password = PHTTPPasswordField::Decrypt(cfg.GetString(PasswordKey, "changeme"));
diff -pruN a/simple/main.cxx b/simple/main.cxx
--- a/simple/main.cxx	2011-03-31 17:12:01.000000000 +0300
+++ b/simple/main.cxx	2011-08-24 14:16:46.000000000 +0300
@@ -907,7 +907,11 @@ PBoolean SimpleH323EndPoint::OpenVideoCh
 
 #ifdef H323_H239
 PBoolean SimpleH323EndPoint::OpenExtendedVideoChannel(H323Connection & connection,
+#if PTRACING
 											PBoolean PTRACE_PARAM(isEncoding),
+#else
+											PBoolean isEncoding,
+#endif
 											H323VideoCodec & codec)
 {
 
