--- a/src/stc/scintilla/src/Editor.cxx	2013-02-27 23:04:10.000000000 +0300
+++ b/src/stc/scintilla/src/Editor.cxx	2016-08-31 15:34:13.688788867 +0300
@@ -16,6 +16,7 @@
 #include <map>
 #include <algorithm>
 #include <memory>
+#include <cmath>
 
 #include "Platform.h"
 
@@ -5841,9 +5842,9 @@ void Editor::GoToLine(int lineNo) {
 }
 
 static bool Close(Point pt1, Point pt2) {
-	if (abs(pt1.x - pt2.x) > 3)
+	if (fabs(pt1.x - pt2.x) > 3)
 		return false;
-	if (abs(pt1.y - pt2.y) > 3)
+	if (fabs(pt1.y - pt2.y) > 3)
 		return false;
 	return true;
 }
