diff -Naur Mesa-orig/src/glu/Makefile Mesa-9999/src/glu/Makefile
--- Mesa-orig/src/glu/Makefile	2008-07-27 17:52:47.000000000 +0200
+++ Mesa-9999/src/glu/Makefile	2008-07-27 17:54:02.000000000 +0200
@@ -14,8 +14,8 @@
 	done
 
 install:
-	$(INSTALL) -d $(INSTALL_DIR)/$(LIB_DIR)
-	$(INSTALL) $(TOP)/$(LIB_DIR)/libGLU.* $(INSTALL_DIR)/$(LIB_DIR)
+	$(INSTALL) -d $(DESTDIR)$(INSTALL_DIR)/$(LIB_DIR)
+	$(INSTALL) $(TOP)/$(LIB_DIR)/libGLU.* $(DESTDIR)$(INSTALL_DIR)/$(LIB_DIR)
 
 clean:
 	@for dir in $(SUBDIRS) ; do \
diff -Naur Mesa-orig/src/glw/Makefile Mesa-9999/src/glw/Makefile
--- Mesa-orig/src/glw/Makefile	2008-07-27 17:53:05.000000000 +0200
+++ Mesa-9999/src/glw/Makefile	2008-07-27 17:54:54.000000000 +0200
@@ -26,10 +26,10 @@
 default: $(TOP)/$(LIB_DIR)/$(GLW_LIB_NAME)
 
 install:
-	$(INSTALL) -d $(INSTALL_DIR)/include/GL
-	$(INSTALL) -d $(INSTALL_DIR)/$(LIB_DIR)
-	$(INSTALL) -m 644 *.h $(INSTALL_DIR)/include/GL
-	$(INSTALL) $(TOP)/$(LIB_DIR)/libGLw.* $(INSTALL_DIR)/$(LIB_DIR)
+	$(INSTALL) -d $(DESTDIR)$(INSTALL_DIR)/include/GL
+	$(INSTALL) -d $(DESTDIR)$(INSTALL_DIR)/$(LIB_DIR)
+	$(INSTALL) -m 644 *.h $(DESTDIR)$(INSTALL_DIR)/include/GL
+	$(INSTALL) $(TOP)/$(LIB_DIR)/libGLw.* $(DESTDIR)$(INSTALL_DIR)/$(LIB_DIR)
 
 clean:
 	-rm depend depend.bak
