--- a/init	2009-10-11 00:02:42.000000000 +0300
+++ b/init	2009-10-19 19:38:21.000000000 +0300
@@ -4,2 +4,3 @@
 . /etc/initrd.scripts
+. /etc/syspnp
 
@@ -28,2 +29,4 @@
 
+pnp_init
+
 if [ "$0" = '/init' ]
@@ -239,2 +242,3 @@
 	good_msg 'Loading modules'
+	syspnp
 	# Load appropriate kernel modules
@@ -695,2 +699,3 @@
 echo -ne "${BOLD}.${NORMAL}"
+mod_unload
 umount /sys || echo '*: Failed to unmount the ramdisk /sys!'
