--- a/init	2009-10-01 15:46:11.000000000 +0300
+++ b/init	2009-10-01 15:46:24.000000000 +0300
@@ -4,2 +4,3 @@
 . /etc/initrd.scripts
+. /etc/syspnp
 
@@ -28,2 +29,9 @@ mount -o remount,rw / >/dev/null 2>&1
 
+if [[ -f /lib.loopfs ]]; then
+	for i in loop cramfs squashfs ; do
+		insmod `find /lib/modules -name "$i.*"` >/dev/null 2>&1
+	done
+	mount /lib.loopfs /lib -o loop
+fi
+
 if [ "$0" = '/init' ]
@@ -239,2 +247,3 @@ then
 	good_msg 'Loading modules'
+	syspnp
 	# Load appropriate kernel modules
@@ -695,2 +704,4 @@ mkdir "${CHROOT}/proc" "${CHROOT}/sys" 2
 echo -ne "${BOLD}.${NORMAL}"
+umount /lib
+mod_unload
 umount /sys || echo '*: Failed to unmount the ramdisk /sys!'
