diff -pruN a/config/udev.c b/config/udev.c
--- a/config/udev.c	2016-08-16 08:11:45.000000000 +0300
+++ b/config/udev.c	2016-11-16 17:19:01.287307830 +0300
@@ -38,6 +38,9 @@
 #include "os.h"
 #include "globals.h"
 #include "systemd-logind.h"
+#ifdef __GLIBC__
+#include <sys/sysmacros.h>
+#endif
 
 #define UDEV_XKB_PROP_KEY "xkb"
 
diff -pruN a/hw/xfree86/common/xf86Xinput.c b/hw/xfree86/common/xf86Xinput.c
--- a/hw/xfree86/common/xf86Xinput.c	2016-10-28 18:09:44.000000000 +0300
+++ b/hw/xfree86/common/xf86Xinput.c	2016-11-16 17:19:52.037305279 +0300
@@ -87,6 +87,9 @@
 #ifdef HAVE_SYS_MKDEV_H
 #include <sys/mkdev.h>          /* for major() & minor() on Solaris */
 #endif
+#ifdef __GLIBC__
+#include <sys/sysmacros.h>
+#endif
 
 #include "mi.h"
 
diff -pruN a/hw/xfree86/os-support/linux/lnx_init.c b/hw/xfree86/os-support/linux/lnx_init.c
--- a/hw/xfree86/os-support/linux/lnx_init.c	2016-08-11 22:08:48.000000000 +0300
+++ b/hw/xfree86/os-support/linux/lnx_init.c	2016-11-16 17:19:36.344306068 +0300
@@ -39,6 +39,10 @@
 
 #include <sys/stat.h>
 
+#ifdef __GLIBC__
+#include <sys/sysmacros.h>
+#endif
+
 #ifndef K_OFF
 #define K_OFF 0x4
 #endif
