OpenMCU SF CVS snapshotted 22.10.2010
MPL-1.0
--- a/conference.h
+++ b/conference.h
@@ -30,6 +30,9 @@
  *                 Craig Southeren (craig@postincrement.com)
  *
  * $Log: conference.h,v $
+ * Revision 1.5  2009/12/23 14:06:57  willamowius
+ * fix signed/unsigned mismatch
+ *
  * Revision 1.4  2009/02/24 11:47:51  willamowius
  * compatibility with older H323Plus versions
  *
@@ -363,7 +366,7 @@
     virtual void CalcVideoSplitSize(unsigned int imageCount, int & subImageWidth, int & subImageHeight, int & cols, int & rows);
     virtual void ReallocatePositions();
     virtual PBoolean ReadMixedFrame(void * buffer, int width, int height, PINDEX & amount);
-    PBoolean ReadSrcFrame(VideoFrameStoreList & srcFrameStores, void * buffer, int width, int height, PINDEX & amount);
+    PBoolean ReadSrcFrame(VideoFrameStoreList & srcFrameStores, void * buffer, unsigned int width, unsigned int height, PINDEX & amount);
 
     PMutex mutex;
     PBoolean forceScreenSplit;
--- a/h323.cxx
+++ b/h323.cxx
@@ -30,6 +30,12 @@
  *                 Craig Southeren (craig@postincrement.com)
  *
  * $Log: h323.cxx,v $
+ * Revision 1.6  2010/03/04 10:01:25  willamowius
+ * don't start MCU if it can't open thge H.323 listener
+ *
+ * Revision 1.5  2010/02/10 09:16:49  shorne
+ * fix for locking and to support remote framesizes other than CIF
+ *
  * Revision 1.4  2008/05/23 11:18:15  willamowius
  * switch BOOL to PBoolean to be able to compile with Ptlib 2.2.x
  *
@@ -207,6 +213,7 @@
 
   if (listeners.IsEmpty()) {
     PSYSTEMLOG(Fatal, "Main\tCould not open H.323 Listener");
+    exit(1);
   }
 
   AliasList.RemoveAll();
@@ -813,7 +820,7 @@
       return FALSE;
     }
 
-    if (!InitGrabber(videoGrabber, CIF_WIDTH, CIF_HEIGHT)) {
+    if (!InitGrabber(videoGrabber, codec.GetWidth(), codec.GetHeight())) {
       delete videoGrabber;
       videoGrabber = NULL;
       return FALSE;
@@ -1182,13 +1189,11 @@
 
 PString H323Connection_ConferenceMember::GetTitle() const
 {
-  PString output;
-  OpenMCUH323Connection * conn = (OpenMCUH323Connection *)ep.FindConnectionWithLock(h323Token);
+  OpenMCUH323Connection * conn = (OpenMCUH323Connection *)GetID(); 
   if (conn != NULL) {
-    output = conn->GetRemoteName();
-    conn->Unlock();
+    return conn->GetRemoteName();
   }
-  return output;
+  return PString();
 }
 
 PString H323Connection_ConferenceMember::GetMonitorInfo(const PString & hdr)
--- a/mcu.cxx
+++ b/mcu.cxx
@@ -30,6 +30,9 @@
  *                 Craig Southeren (craig@postincrement.com)
  *
  * $Log: mcu.cxx,v $
+ * Revision 1.6  2009/08/29 01:54:42  shorne
+ * Fix for changes in PTLIB 2.7
+ *
  * Revision 1.5  2009/05/26 18:16:41  willamowius
  * updated for PTLib 3.6.x
  *
@@ -84,6 +87,7 @@
 
 #include <ptlib.h>
 
+
 #include "version.h"
 #include "mcu.h"
 #include "h323.h"
@@ -92,6 +96,9 @@
 
 extern PHTTPServiceProcess::Info ProductInfo;
 
+static const char ServiceSimulationSectionName[] = "Service Simulation Parameters";
+static const char SystemLogFileNameKey[] = "System Log File Name";
+
 static const char LogLevelKey[]           = "Log Level";
 static const char UserNameKey[]           = "Username";
 static const char PasswordKey[]           = "Password";
@@ -227,19 +234,22 @@
 {
   PConfig cfg("Parameters");
 
-  // Set log level as early as possible
-//  SetLogLevel((PSystemLog::Level)cfg.GetInteger(LogLevelKey, GetLogLevel()));
+#if PTLIB_VER > 260
+   PConfig scfg(ServiceSimulationSectionName);
+   PString systemLogFileName = scfg.GetString(SystemLogFileNameKey);
+#endif
+
 #if PTRACING
+    // Set log level as early as possible
+  SetLogLevel((PSystemLog::Level)cfg.GetInteger(LogLevelKey, GetLogLevel()));
+  //  PTrace::Initialise(6,"c:\\trace.txt");
 
-    SetLogLevel(PSystemLog::Debug6);
-    PTrace::Initialise(6,"c:\\trace.txt");
-
-/*  if (GetLogLevel() >= PSystemLog::Warning)
+  if (GetLogLevel() >= PSystemLog::Warning)
     PTrace::SetLevel(GetLogLevel()-PSystemLog::Warning);
   else
     PTrace::SetLevel(0);
   PTrace::ClearOptions(PTrace::Timestamp);
-  PTrace::SetOptions(PTrace::DateAndTime); */
+  PTrace::SetOptions(PTrace::DateAndTime); 
 #endif
 
   // Get the HTTP basic authentication info
@@ -309,7 +319,7 @@
   // Create invite conference page
   httpNameSpace.AddResource(new InvitePage(*this, authority), PHTTPSpace::Overwrite);
 
-  // Add log file links
+   // Add log file links
   if (!systemLogFileName && (systemLogFileName != "-")) {
     httpNameSpace.AddResource(new PHTTPFile("logfile.txt", systemLogFileName, authority));
     httpNameSpace.AddResource(new PHTTPTailFile("tail_logfile", systemLogFileName, authority));
--- /dev/null
+++ b/openmcu.ini
@@ -0,0 +1,18 @@
+; sample config file for OpenMCU 2.0
+[Parameters]
+;Interface Array Size=1
+;Interface 1=192.168.1.189:1720
+Local User Name=OpenMCU
+
+;Log Level=1
+;Log Level=3
+Log Level=5
+Call log filename=/tmp/openmcu_call.log
+
+Connecting WAV File=connecting.wav
+Entering WAV File=entering.wav
+Leaving WAV File=leaving.wav
+
+Gatekeeper Mode=None
+;Gatekeeper Mode=Find gatekeeper
+
--- a/video.cxx
+++ b/video.cxx
@@ -30,6 +30,9 @@
  *                 Craig Southeren (craig@postincrement.com)
  *
  * $Log: video.cxx,v $
+ * Revision 1.5  2009/12/23 14:06:57  willamowius
+ * fix signed/unsigned mismatch
+ *
  * Revision 1.4  2008/05/23 11:18:19  willamowius
  * switch BOOL to PBoolean to be able to compile with Ptlib 2.2.x
  *
@@ -945,7 +948,7 @@
   return ReadMixedFrame(buffer, width, height, amount);
 }
 
-PBoolean MCUSimpleVideoMixer::ReadSrcFrame(VideoFrameStoreList & srcFrameStores, void * buffer, int width, int height, PINDEX & amount)
+PBoolean MCUSimpleVideoMixer::ReadSrcFrame(VideoFrameStoreList & srcFrameStores, void * buffer, unsigned int width, unsigned int height, PINDEX & amount)
 {
   PWaitAndSignal m(mutex);
 
