--- a/src/nv_driver.c	2011-02-03 19:45:32.000000000 +0200
+++ b/src/nv_driver.c	2011-02-03 19:45:39.000000000 +0200
@@ -795,10 +795,12 @@ NVPreInit(ScrnInfoPtr pScrn, int flags)
 
 	if (!pNv->NoAccel) {
 		if (pNv->Architecture >= NV_ARCH_50)
+		{
 			pNv->wfb_enabled = xf86ReturnOptValBool(
 				pNv->Options, OPTION_WFB, FALSE);
 
 		pNv->tiled_scanout = TRUE;
+		}
 	}
 
 	if (!pNv->NoAccel && pNv->dev->chipset >= 0x11) {
