--- a/src/jemalloc.c	2020-12-21 13:51:52.056920174 +0300
+++ b/src/jemalloc.c	2020-12-21 13:51:05.516918431 +0300
@@ -3596,7 +3596,8 @@ je_sdallocx(void *ptr, size_t size, int
 		size, flags);
 
 	if (flags !=0 || !free_fastpath(ptr, size, true)) {
-		sdallocx_default(ptr, size, flags);
+		//sdallocx_default(ptr, size, flags);
+		je_free(ptr);
 	}
 
 	LOG("core.sdallocx.exit", "");
@@ -3608,7 +3609,8 @@ je_sdallocx_noflags(void *ptr, size_t si
 		size);
 
 	if (!free_fastpath(ptr, size, true)) {
-		sdallocx_default(ptr, size, 0);
+		//sdallocx_default(ptr, size, 0);
+		je_free(ptr);
 	}
 
 	LOG("core.sdallocx.exit", "");
