--- a/src/jemalloc_cpp.cpp	2020-12-21 11:05:33.906805209 +0300
+++ b/src/jemalloc_cpp.cpp	2020-12-21 11:07:15.010142330 +0300
@@ -24,6 +24,9 @@ extern "C" {
 //
 // ... but it needs to work with jemalloc namespaces.
 
+#define je_sdallocx_noflags(ptr,size) if (!free_fastpath(ptr, size, true)) je_free(ptr)
+#define je_sdallocx(ptr,size,flags) if (flags != 0 || !free_fastpath(ptr, size, true)) je_free(ptr)
+
 void	*operator new(std::size_t size);
 void	*operator new[](std::size_t size);
 void	*operator new(std::size_t size, const std::nothrow_t &) noexcept;
